### PR TITLE
Versioneer produces version on building

### LINKED ISSUE #107: Automate release of tagged versions to PyPi and ORDA
We have the package on [PyPi](https://pypi.org/) but we would like to automate releases and updates when a tagged version is released.

Similarly we would like to automate the release of packages to [ORDA](https://orda.sheffield.ac.uk/) (which uses Figshare).

* [pypi-publish](https://github.com/marketplace/actions/pypi-publish)
* [fishare/github-upload-action](https://github.com/figshare/github-upload-action)

See also @bobturneruk very useful notes as per #72 on how to [Release to ORDA](https://github.com/RSE-Sheffield/release_to_ORDA)

### PR DESCRIPTION
**Pull request by ns-rse**
Resolves #107

This PR introduces [versioneer](https://github.com/python-versioneer/python-versioneer) which automatically determines the version of a package when building it for release.

This is required so that we can automate the releasing of packages to PyPI and ORDA. The GitHub Actions (under `.github/workflows/pypi.yaml` and `.github/workflows/ORDA.yaml`) should then automatically be triggered to push packages to PyPI and ORDA respectively when a Git commit is tagged (on GitHub or locally and pushed), resulting in updates to both.

### CODE REVIEW

**Comment by ns-rse** [comment]
After much faffing and scratching of head I corrected a typo in the configuration of `versioneer` that prevent valid PEP440 tags from being created.

I've also now understood more about how the `style` of the tags and by using `style = pep440-pre` was able to successfully push to `test.pypi.org` (see [here](https://test.pypi.org/project/pgfinder/0.0.1.post2.dev3/)).

Its important to note that `versioneer` produces tags that are non-compliant with PEP440 if the commit is ahead of a tag. This means that it shouldn't be a problem using `style = pep440` since the `.github/workflow/pypi.yaml` GitHub Action should _only_ be run when a commit has been tagged (i.e. there shouldn't be any commits after the tag).

**Review by bobturneruk — APPROVED** [review]
Looks good. 

It's a shame to have to vendor in so much code to make this happen, but it's helpful automation.  Glad you've figure this all out!

Just in case I've misunderstood: The package version is now not specified anywhere in the code or metadata, but extracted from the git history when the package is built (or dev installed from a git repo.). Might be good to note this in the docs.

**Comment by bobturneruk** [comment]
Also, can we exclude `_version.py` from test coverage reporting somehow?

File: .github/workflows/ci-tests.yml [hunk 1]
```diff
@@ -21,8 +21,10 @@ jobs:
           python-version: ${{ matrix.python-version }}
       - name: Install dependencies
         run: |
+          pip3 install git+https://github.com/python-versioneer/python-versioneer@0337a75f1cc781f4a7ea35baab9545ae964b0ec7
```
[the remark this answers is not in the document]
Comment (ns-rse) [reply .github/workflows/ci-tests.yml:24]: Thanks, forgotten I'd added this and can't recall why `versioneer` was required to be installed for the tests.

File: .github/workflows/pypi.yaml [hunk 2]
```diff
@@ -19,12 +19,13 @@ jobs:
           python-version: 3.9
       - name: Installing the package
         run: |
+          pip3 install versioneer
           pip3 install .
           pip3 install .[pypi]
```
[the remark this answers is not in the document]
Comment (ns-rse) [reply .github/workflows/pypi.yaml:24]: There was a problem with versioneer-0.24 when I was originally doing this (see [versioneer doesn't work on recent pip when PEP 517 build is triggered](https://github.com/python-versioneer/python-versioneer/issues/192)) and so I had originally installed from a commit that fixed it (code left in other comment) and wasn't paying attention when updating.

Removed so it should be pulled in with `pip3 install .[pypi]`.

**Comment by ns-rse** [comment]
> Just in case I've misunderstood: The package version is now not specified anywhere in the code or metadata, but extracted from the git history when the package is built (or dev installed from a git repo.). Might be good to note this in the docs.

Nope, you've got it, versioneer pulls a version from the last tag. If `HEAD` has moved ahead of this then the current commit is included along with the distance from the last tag. Good point about noting this in the documentation I've added it.

> Also, can we exclude _version.py from test coverage reporting somehow?

It should already be since the invocation in `.github/workflows/ci-tests.yaml` is...

```
pytest --cov=pgfinder
```

However, this may have been being over-ridden by the configuration present in `pyproject.toml` which had `addopts = "--cov"` for pytest, I've udpated that to `addopts = "--cov=pgfinder"`.

**Pull request merged**